modules/core/src/main/java/org/apache/ignite/internal/processors/cache/PartitionUpdateCounter.java

```diff
@@ -180,14 +181,24 @@ private void offer(Item item) {
 
     /**
      * Flushes pending update counters closing all possible gaps.
+     *
+     * @param onGapClose Closure to run on gap closing.
      */
-    public synchronized void finalizeUpdateCountres() {
-        Item last = queue.pollLast();
+    public synchronized void finalizeUpdateCounters(IgniteInClosure<Long> onGapClose) {
+        Item item = poll();
+
+        while (item != null) {
+            long cur = cntr.get();
 
-        if (last != null)
-            update(last.start + last.delta);
+            // Call closure for each counter between current value and next range start.
+            for (long cntr0 = cur; cntr0 < item.start; cntr0++)
+                onGapClose.apply(cntr0 + 1);
 
-        queue.clear();
+            // Close pending ranges.
+            update(item.start + item.delta);
```

Comment:
Can we call update only for the last item?

Reply:
No, we can't. We need too iterate over all items in the queue and handle all gaps between them.

Reply:
I meant `update` method specifically. As far as I see we need to call `update` only for the greatest interval. I could suggest something like this:
```
if (queue.isEmpty())
    return;

long cur = cntr.get();

for (Item item : queue) {
    for (long c = cur + 1; c <= item.start; c++)
        onGapClose.apply(c);

    cur = item.start;
}

Item last = queue.pollLast();

update(last.start + last.delta);

queue.clear();
```

Reply:
Yes, your variant is slightly better from the performance point of view. But this code is not on the hot path and even more - this is a very, very cold path. Gaps happens in the very rare cases. So it is not the best place to start optimization from.